### PR TITLE
Split out Bag theory, elaborate Maps and Bags into Arrays

### PR DESCRIPTION
**Pull request by clayrat**
As a continuation of https://github.com/ucsd-progsys/liquid-fixpoint/pull/688, this PR elaborates Map operations into Array ones, allowing for the correct handling of polymorphic Map values.

It also separates Bags (i.e., integer-valued maps) from generic Maps and removes some ad-hoc Map operations that are currently only implementable as Z3 lambdas (`prj`, `shift` and `to_set`). This is a precursor to adding CVC5 support, as the latter does not support arbitrary Array operations, but instead has a generic [SMTLIB Array theory](https://smt-lib.org/theories-ArraysEx.shtml) (with `const`, `store` and `select`) and separate theories of [Sets](https://cvc5.github.io/docs-ci/docs-main/theories/sets-and-relations.html) and [Bags](https://cvc5.github.io/docs-ci/docs-main/theories/bags.html).

### CODE REVIEW

File: tests/neg/maps.fq [hunk 1]
```diff
@@ -1,37 +1,30 @@
...
   env [ 1; 2; 3 ]
   lhs {v : int | true }
-  rhs {v : int | m2 = m3 } 
+  rhs {v : int | m4 = m5 }
   id 4 tag []
-
-constraint:
-  env [ 1; 2; 3; 4; 5 ]
-  lhs {v : int | true }
-  rhs {v : int | m2 = Map_union m4 m4 }
```
Comment (nikivazou) [inline tests/neg/maps.fq:35]: This is removed because Map_union is no more supported?

**Review by nikivazou — APPROVED** [review]
Looks good to me!

File: tests/neg/maps.fq (hunk 1 above)
Comment (clayrat) [reply tests/neg/maps.fq:35]: Yes, this operation is only defined for Maps with integer values, so it got moved into the Bag theory as `Bag_union`

**Comment by clayrat** [comment]
For the reference, the Map operations we're removing here were first introduced in https://github.com/ucsd-progsys/liquid-fixpoint/pull/624

**Pull request merged**